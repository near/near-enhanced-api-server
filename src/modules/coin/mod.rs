use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
