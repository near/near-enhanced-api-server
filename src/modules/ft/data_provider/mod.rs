mod balance;
mod history;
mod metadata;
mod models;

pub(crate) use balance::{get_ft_balance_by_contract, get_ft_balances};
pub(crate) use history::get_ft_history;
pub(crate) use metadata::get_ft_metadata;
