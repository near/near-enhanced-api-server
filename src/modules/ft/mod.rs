use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    // I'll drop these comments, leaving it only because it's easier for the review
    // app.service(
    //     web::resource("/accounts/{account_id}/NEAR")
    //         .route(web::get().to(resources::get_near_balance)),
    // )
    app.service(
        web::resource("/accounts/{account_id}/FT").route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    // .service(
    //     web::resource("/accounts/{account_id}/NEAR/history")
    //         .route(web::get().to(resources::get_near_history)),
    // )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
