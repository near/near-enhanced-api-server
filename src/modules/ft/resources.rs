use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(FT))]
/// Get user's FT balances
///
/// This endpoint returns all non-zero FT balances of the given `account_id`,
/// at the given `block_timestamp_nanos`/`block_height`.
///
/// **Limitations**
/// This endpoint scans all the FT contracts.
/// We currently provide up to 100 results, which covers almost all the potential situations.
/// Anyway, full-featured pagination will be provided soon.
pub async fn get_ft_balances(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    pool_balances: web::Data<db_helpers::BalancesPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    // TODO pagination
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::FtBalancesResponse>> {
    let limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool_explorer, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_ft_balances(
        &pool_balances,
        &rpc_client,
        &request.account_id.0,
        &block,
        limit,
    )
    .await?;

    Ok(Json(schemas::FtBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get user's FT balance by contract
///
/// This endpoint returns FT balance of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
pub async fn get_ft_balance_by_contract(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtBalanceByContractResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use `/accounts/{account_id}/balances/NEAR`".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool_explorer, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balance = data_provider::get_ft_balance_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::FtBalanceByContractResponse {
        balance,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get user's FT history by contract
///
/// This endpoint returns the history of FT operations
/// for the given `account_id`, `contract_account_id`.
/// For the next page, use `event_index` of the last item in your previous response.
pub async fn get_ft_history(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    pool_balances: web::Data<db_helpers::BalancesPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::FtHistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use `/accounts/{account_id}/balances/NEAR/history`"
                .to_string(),
        )
        .into());
    }
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::get_block_from_pagination(&pool_explorer, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::FtHistoryResponse {
        history: data_provider::get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get FT metadata
///
/// This endpoint returns the metadata for the given `contract_account_id`, `block_timestamp_nanos`/`block_height`.
pub async fn get_ft_metadata(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool_explorer, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
