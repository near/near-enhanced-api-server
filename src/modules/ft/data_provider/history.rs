use crate::modules::ft;
use crate::{db_helpers, errors, types};
use num_traits::{Signed, ToPrimitive};
use sqlx::types::BigDecimal;
use std::str::FromStr;

pub(crate) async fn get_ft_history(
    db_helpers::ExplorerPool(pool_explorer): &db_helpers::ExplorerPool,
    db_helpers::BalancesPool(pool_balances): &db_helpers::BalancesPool,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
    block: &db_helpers::Block,
    pagination: &types::query_params::Pagination,
) -> crate::Result<Vec<ft::schemas::HistoryItem>> {
    let metadata = ft::schemas::Metadata::from(
        super::metadata::get_ft_metadata(rpc_client, contract_id.clone(), block.height).await?,
    );

    let after_event_index = if let Some(index) = pagination.after_event_index {
        index
    } else {
        // +1 because we need to include given timestamp to result. Query has strict less operator
        db_helpers::timestamp_to_event_index(block.timestamp + 1)
    };

    // We don't have absolute_value in the DB.
    // We ask RPC for the first absolute_value, fill the in-between values with deltas, check by asking RPC for the last absolute_value and comparing the results.
    // If it matches, everything is good. If not, we return the error.

    // Problem: we can ask RPC only for the balance at the end of the block, while we may have the lines starting and ending in the middle of the block.
    // That's why we need to take slightly more lines so that we can match it with the RPC, and then filter these lines.
    // Potential issue here: the DB may contain only the part of the most fresh block (we may be in the middle of the writing process)
    // It can be easily solved by ignoring the most fresh block, but it will increase the lag between the response and the current blockchain state.
    // Since we anyway use transactional DB which guarantees that write process goes atomically, I don't want to do anything with that.
    // But, if we meet such issues in production, we may consider cutting the latest block.
    // TODO check the performance. We may add index on block_timestamp column, or we can hack and change block_timestamp to event_index
    let query = r"
         WITH original_query as (
             SELECT
                 event_index,
                 involved_account_id,
                 delta_amount,
                 cause,
                 status,
                 block_timestamp,
                 block_height
             FROM coin_events
             WHERE contract_account_id = $1
                 AND affected_account_id = $2
                 AND event_index < $3::numeric(38, 0)
                 AND event_index >= 16619903149406361800000000000000000 -- todo drop this when we finish collecting the data
             ORDER BY event_index desc
             LIMIT $4::numeric(20, 0)
         ), timestamps as (
             SELECT
                 min(block_timestamp) min_block_timestamp,
                 max(block_timestamp) max_block_timestamp
             FROM original_query
         )
         SELECT
             event_index,
             involved_account_id,
             delta_amount delta_balance,
             cause,
             status,
             block_timestamp block_timestamp_nanos,
             block_height
         FROM coin_events, timestamps
         WHERE contract_account_id = $1
             AND affected_account_id = $2
             AND block_timestamp >= min_block_timestamp
             AND block_timestamp <= max_block_timestamp
         ORDER BY event_index desc
     ";

    let history = db_helpers::select_retry_or_panic::<ft::data_provider::models::FtHistoryInfo>(
        pool_balances,
        query,
        &[
            contract_id.to_string(),
            account_id.to_string(),
            after_event_index.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut current_balance = if let Some(first_item) = history.first() {
        let amount = super::balance::get_ft_amount(
            rpc_client,
            contract_id.clone(),
            account_id.clone(),
            first_item.block_height.to_u64().ok_or_else(|| {
                errors::ErrorKind::InternalError(
                    "Found negative block_height in coin_events table".to_string(),
                )
            })?,
        )
        .await?;
        BigDecimal::from_str(&amount.to_string()).map_err(|e| {
            errors::ErrorKind::InternalError(format!("Failed to parse BigDecimal from u128: {}", e))
        })?
    } else {
        return Ok(vec![]);
    };

    let mut result: Vec<ft::schemas::HistoryItem> = vec![];
    for db_info in history {
        let balance = current_balance.clone();

        if db_info.status == "SUCCESS" {
            current_balance -= db_info.delta_balance.clone();
            if current_balance.is_negative() {
                tracing::warn!(
                    target: crate::LOGGER_MSG,
                    "get_ft_history: found inconsistent events for contract {}, account {}, block {:#?}, pagination {:#?}",
                    contract_id,
                    account_id,
                    block,
                    pagination,
                );
                return Err(errors::ErrorKind::InternalError(format!(
                    "History is not supported for account {}. Contract {} provides inconsistent events which lead to negative balance",
                    account_id, contract_id
                ))
                    .into());
            }
        }

        let involved_account_id = match db_info.involved_account_id {
            Some(id) => Some(types::AccountId::from_str(&id)?),
            None => None,
        };
        let event_index = types::numeric::to_u128(&db_info.event_index)?;

        // We collect slightly more lines that we were asked for, because we can make RPC calls only at the end of the block
        // First clause filters latest redundant lines, second clause filters earliest redundant lines
        if pagination
            .after_event_index
            .map_or_else(|| true, |index| index > event_index)
            && result.len() < pagination.limit as usize
        {
            result.push(ft::schemas::HistoryItem {
                event_index: event_index.into(),
                cause: db_info.cause.clone(),
                involved_account_id,
                delta_balance: db_info.delta_balance.to_string(),
                balance: types::numeric::to_u128(&balance)?.into(),
                block_timestamp_nanos: types::numeric::to_u64(&db_info.block_timestamp_nanos)?
                    .into(),
                block_height: types::numeric::to_u64(&db_info.block_height)?.into(),
                status: db_info.status,
                metadata: metadata.clone(),
            });
        }
    }

    let prev_block = if let Some(item) = result.last() {
        db_helpers::get_previous_block(pool_explorer, item.block_timestamp_nanos.0).await?
    } else {
        return Ok(result);
    };
    let earliest_balance = match super::balance::get_ft_amount(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        prev_block.height,
    )
    .await
    {
        Ok(x) => x,
        Err(e) => {
            if e.message.contains("does not exist at block_height") {
                0
            } else {
                return Err(e);
            }
        }
    };
    if types::numeric::to_u128(&current_balance)? != earliest_balance {
        tracing::warn!(
            target: crate::LOGGER_MSG,
            "get_ft_history: found inconsistent events for contract {}, account {}, block {:#?}, pagination {:#?}",
            contract_id,
            account_id,
            block,
            pagination,
        );
        return Err(errors::ErrorKind::InternalError(format!(
            "History is not supported for account {}. Contract {} provides inconsistent events",
            account_id, contract_id
        ))
        .into());
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;
    use std::str::FromStr;

    #[tokio::test]
    async fn test_ft_history() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str(
            "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
        )
        .unwrap();
        let account = near_primitives::types::AccountId::from_str("aurora").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: None,
        };

        let balance = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_ft_history_next_page() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let contract = near_primitives::types::AccountId::from_str(
            "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
        )
        .unwrap();
        let account = near_primitives::types::AccountId::from_str("aurora").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: Some(16708552830626965310000000004000001),
        };
        let block = db_helpers::get_block_from_pagination(&pool_explorer, &pagination)
            .await
            .unwrap();

        let history = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(history);
    }

    #[tokio::test]
    async fn test_ft_history_next_page_in_the_middle_of_the_block() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let contract = near_primitives::types::AccountId::from_str(
            "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
        )
        .unwrap();
        let account = near_primitives::types::AccountId::from_str("aurora").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: Some(16704039164216566310000000004000001),
        };
        let block = db_helpers::get_block_from_pagination(&pool_explorer, &pagination)
            .await
            .unwrap();

        let history1 = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await
        .unwrap();

        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: Some(history1.last().unwrap().event_index.0),
        };
        let block = db_helpers::get_block_from_pagination(&pool_explorer, &pagination)
            .await
            .unwrap();
        let history2 = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await
        .unwrap();

        assert!(
            history1.last().unwrap().event_index > history2.first().unwrap().event_index,
            "Next page should not include event from previous page"
        );
        assert_eq!(
            history1.last().unwrap().block_height,
            history2.first().unwrap().block_height,
            "Block split in the middle expected"
        );
    }

    #[tokio::test]
    async fn test_ft_history_with_failed_receipts() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str(
            "52a047ee205701895ee06a375492490ec9c597ce.factory.bridge.near",
        )
        .unwrap();
        let account = near_primitives::types::AccountId::from_str("v2.ref-finance.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: None,
        };

        let balance = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_ft_history_account_never_existed() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account =
            near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let history = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await
        .unwrap();
        assert!(history.is_empty());
    }

    #[tokio::test]
    async fn test_ft_history_account_deleted() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account =
            near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let history = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await
        .unwrap();
        assert!(history.is_empty());
    }

    #[tokio::test]
    async fn test_ft_history_contract_does_not_exist() {
        let pool_explorer = init_explorer_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract =
            near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("olga.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let balance = get_ft_history(
            &pool_explorer,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(balance);
    }

    // TODO write this test again when we finish collecting the data
    // I can't catch such cases on partially filled DB
    // #[tokio::test]
    // async fn test_ft_history_contract_is_inconsistent() {
    //     let pool_explorer = init_db().await;
    //     let pool_balances = init_balances_db().await;
    //     let rpc_client = init_rpc();
    //
    //     let contract = near_primitives::types::AccountId::from_str("tezeract.near").unwrap();
    //     let account = near_primitives::types::AccountId::from_str("puffball.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 5,
    //         after_event_index: Some(16629459979548196140000003001000001),
    //     };
    //     let block = db_helpers::checked_get_block_from_pagination(&pool_explorer, &pagination)
    //         .await
    //         .unwrap();
    //
    //     let balance = get_ft_history(
    //         &pool_explorer,
    //         &pool_balances,
    //         &rpc_client,
    //         &contract,
    //         &account,
    //         &block,
    //         &pagination,
    //     )
    //     .await;
    //     insta::assert_debug_snapshot!(balance);
    // }
}
