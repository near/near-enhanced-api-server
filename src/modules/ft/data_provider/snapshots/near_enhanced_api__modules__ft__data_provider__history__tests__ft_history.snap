---
source: src/modules/ft/data_provider/history.rs
expression: balance
---
Ok(
    [
        HistoryItem {
            event_index: U128(
                16708676171606215200000000004000000,
            ),
            involved_account_id: Some(
                AccountId(
                    "e4090f0e36dbea21cd5a8aa219f9c0b7b07a19563cc48f3eda7110265a6889ca",
                ),
            ),
            delta_balance: "-200000000000000000000",
            balance: U128(
                37661730653372875230123282,
            ),
            cause: "TRANSFER",
            status: "SUCCESS",
            metadata: Metadata {
                name: "Aurora",
                symbol: "AURORA",
                icon: Some(
                    "data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 288 288' style='enable-background:new 0 0 288 288;' xml:space='preserve'%3E %3Cstyle type='text/css'%3E .st0%7Bfill:%2370D44B;%7D .st1%7Bfill:%23FFFFFF;%7D %3C/style%3E %3Cpath class='st0' d='M144,0L144,0c79.5,0,144,64.5,144,144v0c0,79.5-64.5,144-144,144h0C64.5,288,0,223.5,0,144v0 C0,64.5,64.5,0,144,0z'/%3E %3Cpath class='st1' d='M144,58.8c7.6,0,14.5,4.3,17.9,11.1l56.2,112.5c4.9,9.9,0.9,21.9-9,26.8c-2.8,1.4-5.8,2.1-8.9,2.1H87.8 c-11,0-20-9-20-20c0-3.1,0.7-6.2,2.1-8.9l56.2-112.5C129.5,63,136.4,58.7,144,58.8 M144,45c-12.8,0-24.5,7.2-30.2,18.7L57.6,176.2 c-8.3,16.7-1.6,36.9,15.1,45.3c4.7,2.3,9.9,3.6,15.1,3.6h112.5c18.6,0,33.8-15.1,33.8-33.7c0-5.2-1.2-10.4-3.6-15.1L174.2,63.7 C168.5,52.2,156.8,45,144,45z'/%3E %3C/svg%3E",
                ),
                decimals: 18,
            },
            block_timestamp_nanos: U64(
                1670867617160621520,
            ),
            block_height: U64(
                80499933,
            ),
        },
        HistoryItem {
            event_index: U128(
                16708640769464935820000000004000000,
            ),
            involved_account_id: Some(
                AccountId(
                    "e4090f0e36dbea21cd5a8aa219f9c0b7b07a19563cc48f3eda7110265a6889ca",
                ),
            ),
            delta_balance: "-200000000000000000000",
            balance: U128(
                37661930653372875230123282,
            ),
            cause: "TRANSFER",
            status: "SUCCESS",
            metadata: Metadata {
                name: "Aurora",
                symbol: "AURORA",
                icon: Some(
                    "data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 288 288' style='enable-background:new 0 0 288 288;' xml:space='preserve'%3E %3Cstyle type='text/css'%3E .st0%7Bfill:%2370D44B;%7D .st1%7Bfill:%23FFFFFF;%7D %3C/style%3E %3Cpath class='st0' d='M144,0L144,0c79.5,0,144,64.5,144,144v0c0,79.5-64.5,144-144,144h0C64.5,288,0,223.5,0,144v0 C0,64.5,64.5,0,144,0z'/%3E %3Cpath class='st1' d='M144,58.8c7.6,0,14.5,4.3,17.9,11.1l56.2,112.5c4.9,9.9,0.9,21.9-9,26.8c-2.8,1.4-5.8,2.1-8.9,2.1H87.8 c-11,0-20-9-20-20c0-3.1,0.7-6.2,2.1-8.9l56.2-112.5C129.5,63,136.4,58.7,144,58.8 M144,45c-12.8,0-24.5,7.2-30.2,18.7L57.6,176.2 c-8.3,16.7-1.6,36.9,15.1,45.3c4.7,2.3,9.9,3.6,15.1,3.6h112.5c18.6,0,33.8-15.1,33.8-33.7c0-5.2-1.2-10.4-3.6-15.1L174.2,63.7 C168.5,52.2,156.8,45,144,45z'/%3E %3C/svg%3E",
                ),
                decimals: 18,
            },
            block_timestamp_nanos: U64(
                1670864076946493582,
            ),
            block_height: U64(
                80496901,
            ),
        },
        HistoryItem {
            event_index: U128(
                16708630550524950520000000004000001,
            ),
            involved_account_id: Some(
                AccountId(
                    "avan.near",
                ),
            ),
            delta_balance: "14000000000000000000",
            balance: U128(
                37662130653372875230123282,
            ),
            cause: "TRANSFER",
            status: "SUCCESS",
            metadata: Metadata {
                name: "Aurora",
                symbol: "AURORA",
                icon: Some(
                    "data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 288 288' style='enable-background:new 0 0 288 288;' xml:space='preserve'%3E %3Cstyle type='text/css'%3E .st0%7Bfill:%2370D44B;%7D .st1%7Bfill:%23FFFFFF;%7D %3C/style%3E %3Cpath class='st0' d='M144,0L144,0c79.5,0,144,64.5,144,144v0c0,79.5-64.5,144-144,144h0C64.5,288,0,223.5,0,144v0 C0,64.5,64.5,0,144,0z'/%3E %3Cpath class='st1' d='M144,58.8c7.6,0,14.5,4.3,17.9,11.1l56.2,112.5c4.9,9.9,0.9,21.9-9,26.8c-2.8,1.4-5.8,2.1-8.9,2.1H87.8 c-11,0-20-9-20-20c0-3.1,0.7-6.2,2.1-8.9l56.2-112.5C129.5,63,136.4,58.7,144,58.8 M144,45c-12.8,0-24.5,7.2-30.2,18.7L57.6,176.2 c-8.3,16.7-1.6,36.9,15.1,45.3c4.7,2.3,9.9,3.6,15.1,3.6h112.5c18.6,0,33.8-15.1,33.8-33.7c0-5.2-1.2-10.4-3.6-15.1L174.2,63.7 C168.5,52.2,156.8,45,144,45z'/%3E %3C/svg%3E",
                ),
                decimals: 18,
            },
            block_timestamp_nanos: U64(
                1670863055052495052,
            ),
            block_height: U64(
                80496007,
            ),
        },
        HistoryItem {
            event_index: U128(
                16708552840617523550000000004000001,
            ),
            involved_account_id: Some(
                AccountId(
                    "boobs.zudwa.near",
                ),
            ),
            delta_balance: "141281972069835481089",
            balance: U128(
                37662116653372875230123282,
            ),
            cause: "TRANSFER",
            status: "SUCCESS",
            metadata: Metadata {
                name: "Aurora",
                symbol: "AURORA",
                icon: Some(
                    "data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 288 288' style='enable-background:new 0 0 288 288;' xml:space='preserve'%3E %3Cstyle type='text/css'%3E .st0%7Bfill:%2370D44B;%7D .st1%7Bfill:%23FFFFFF;%7D %3C/style%3E %3Cpath class='st0' d='M144,0L144,0c79.5,0,144,64.5,144,144v0c0,79.5-64.5,144-144,144h0C64.5,288,0,223.5,0,144v0 C0,64.5,64.5,0,144,0z'/%3E %3Cpath class='st1' d='M144,58.8c7.6,0,14.5,4.3,17.9,11.1l56.2,112.5c4.9,9.9,0.9,21.9-9,26.8c-2.8,1.4-5.8,2.1-8.9,2.1H87.8 c-11,0-20-9-20-20c0-3.1,0.7-6.2,2.1-8.9l56.2-112.5C129.5,63,136.4,58.7,144,58.8 M144,45c-12.8,0-24.5,7.2-30.2,18.7L57.6,176.2 c-8.3,16.7-1.6,36.9,15.1,45.3c4.7,2.3,9.9,3.6,15.1,3.6h112.5c18.6,0,33.8-15.1,33.8-33.7c0-5.2-1.2-10.4-3.6-15.1L174.2,63.7 C168.5,52.2,156.8,45,144,45z'/%3E %3C/svg%3E",
                ),
                decimals: 18,
            },
            block_timestamp_nanos: U64(
                1670855284061752355,
            ),
            block_height: U64(
                80489370,
            ),
        },
        HistoryItem {
            event_index: U128(
                16708552830626965310000000004000001,
            ),
            involved_account_id: Some(
                AccountId(
                    "resetme.near",
                ),
            ),
            delta_balance: "146097628000000000000",
            balance: U128(
                37661975371400805394642193,
            ),
            cause: "TRANSFER",
            status: "SUCCESS",
            metadata: Metadata {
                name: "Aurora",
                symbol: "AURORA",
                icon: Some(
                    "data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 288 288' style='enable-background:new 0 0 288 288;' xml:space='preserve'%3E %3Cstyle type='text/css'%3E .st0%7Bfill:%2370D44B;%7D .st1%7Bfill:%23FFFFFF;%7D %3C/style%3E %3Cpath class='st0' d='M144,0L144,0c79.5,0,144,64.5,144,144v0c0,79.5-64.5,144-144,144h0C64.5,288,0,223.5,0,144v0 C0,64.5,64.5,0,144,0z'/%3E %3Cpath class='st1' d='M144,58.8c7.6,0,14.5,4.3,17.9,11.1l56.2,112.5c4.9,9.9,0.9,21.9-9,26.8c-2.8,1.4-5.8,2.1-8.9,2.1H87.8 c-11,0-20-9-20-20c0-3.1,0.7-6.2,2.1-8.9l56.2-112.5C129.5,63,136.4,58.7,144,58.8 M144,45c-12.8,0-24.5,7.2-30.2,18.7L57.6,176.2 c-8.3,16.7-1.6,36.9,15.1,45.3c4.7,2.3,9.9,3.6,15.1,3.6h112.5c18.6,0,33.8-15.1,33.8-33.7c0-5.2-1.2-10.4-3.6-15.1L174.2,63.7 C168.5,52.2,156.8,45,144,45z'/%3E %3C/svg%3E",
                ),
                decimals: 18,
            },
            block_timestamp_nanos: U64(
                1670855283062696531,
            ),
            block_height: U64(
                80489369,
            ),
        },
    ],
)
